### PR TITLE
8302659: Modernize Windows native code for NetworkInterface

### PR DESCRIPTION
**Pull request by rdicroce**
Improves performance and correctness, as discussed on the net-dev mailing list.

---------
### Progress
- [x] Change requires CSR request [JDK-8303477](https://bugs.openjdk.org/browse/JDK-8303477) to be approved
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issues
 * [JDK-8302659](https://bugs.openjdk.org/browse/JDK-8302659): Modernize Windows native code for NetworkInterface
 * [JDK-8303477](https://bugs.openjdk.org/browse/JDK-8303477): Modernize Windows native code for NetworkInterface (**CSR**)

### Reviewers
 * [Magnus Ihse Bursie](https://openjdk.org/census#ihse) (@magicus - **Reviewer**) ⚠️ Review applies to [311cdc22](https://git.openjdk.org/jdk/pull/12593/files/311cdc22ddbc23d72912460bc607ac87646f5955)
 * [Daniel Jeliński](https://openjdk.org/census#djelinski) (@djelinski - Committer)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)
 * [Michael McMahon](https://openjdk.org/census#michaelm) (@Michael-Mc-Mahon - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12593/head:pull/12593` \
`$ git checkout pull/12593`

Update a local copy of the PR: \
`$ git checkout pull/12593` \
`$ git pull https://git.openjdk.org/jdk pull/12593/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12593`

View PR using the GUI difftool: \
`$ git pr show -t 12593`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12593.diff">https://git.openjdk.org/jdk/pull/12593.diff</a>

</details>

### CODE REVIEW

**Comment by dfuch** [comment]
FWIW - there is a perl script located in `make/scripts/normalizer.pl` that can be run on modified sources to fix up whitespace  and CRLF issues when jcheck complains.

**Comment by AlanBateman** [comment]
Would it be possible to re-format the changes to src/java.base/windows/native/libnet/NetworkInterface.c to avoid the overly long lines - some of the new lines are 150-200 characters long so it will make impossible to look at side-by-side diffs in the future.

**Comment by rdicroce** [comment]
Sure. What do you want the max line length to be? 80 chars? 100?

**Comment by dfuch** [comment]
Something between 80 and 100 is usually a good limit. We typically avoid to go above that. So no hard limit at 80 but avoid having lines which are too long. If it displays correctly without wrapping using side diff in webrev / github side diff file view it's a good sign.

**Review by magicus — APPROVED** [review]
Build changes look fine.

/reviewers 2

**Comment by rdicroce** [comment]
Maximum line length reduced to 80-ish characters.

File: src/java.base/windows/native/libnet/NetworkInterface.c
Comment (djelinski) [inline src/java.base/windows/native/libnet/NetworkInterface.c]: I'm not a big fan of using WBEM here. Also, I'm still trying to figure out when (if ever) this method is supposed to return false

**Review by djelinski — COMMENTED** [review]
Overall the changes look good. A few comments below.

File: src/java.base/windows/native/libnet/NetworkInterface.c
Comment (rdicroce) [reply src/java.base/windows/native/libnet/NetworkInterface.c:73]: For the IPv6 code, this should be the same as before. See getAddrsFromAdapter in the original code, which enumerates the unicast addresses first and then does the anycast addresses.

For IPv4, I'm not sure. The MS docs for GetIpAddrTable don't say, and I don't have any systems where I could determine it experimentally.

**Comment by msheppar** [comment]
what versions of Windows has this been tested on ? Just wondering if the deletion of the old dual stack winXP based code could have on the older windows OS platforms?
afaik current set of supported windows OS platforms, from the ATR list, are : 
            windows server 2019,
            windows server 2022,
            windows server  2016,
            windows 10,
            windows 11,
            windows 8,
            windows server2012,
            windows server 2012R2

some of these may be retired for LTS JDK21, but for the moment they are in the mix.
We'll need to schedule test runs for all these.

Also we will have to arrange some IPv6 only test runs for this prior to any integration -- that takes a bit of time to arrange the resources to be setup and to schedule the runs.

File: src/java.base/windows/native/libnet/NetworkInterface.c
Comment (rdicroce) [reply src/java.base/windows/native/libnet/NetworkInterface.c]: You and me both. As I wrote in the giant comment, I did a lot of experimenting. In the end, what I determined was that if the IGMPLevel (IPv4) or MldLevel (IPv6) is set to 0 (None), then attempting to join a multicast group throws an exception. So that seemed like a reasonable thing that supportsMulticast() could measure. Nothing else I tried, including the existing GetAdaptersAddresses approach, indicated that multicast was disabled in those situations.

Having said that, I'd be okay with ditching all of this code and just returning true unconditionally. It's unlikely that multicast would be disabled, and a single boolean isn't enough to express the actual setting anyway, since Windows apparently lets you disable receiving multicasts while still allowing you to send them.

**Comment by rdicroce** [comment]
@msheppar I tested this on Windows 10 and 11. I don't have access to any older systems. But I tried to be careful about which APIs I used. I'm pretty sure I checked the MS docs for all of them, and they all exist since Vista/Server 2008. So this should work on all the versions you listed.

File: src/java.base/windows/native/libnet/NetworkInterface.c
Comment (djelinski) [reply src/java.base/windows/native/libnet/NetworkInterface.c]: Ha! Apparently I have a machine where the implementation of `supportsMulticast` implementation makes a difference. On my machine the test `jdk/java/net/SocketOption/OptionsTest.java` is failing with:
```
java.net.SocketException: Invalid argument: no further information
	at java.base/sun.nio.ch.Net.setInterface6(Native Method)
	at java.base/sun.nio.ch.DatagramChannelImpl.setOption(DatagramChannelImpl.java:377)
	at java.base/sun.nio.ch.DatagramSocketAdaptor.setOption(DatagramSocketAdaptor.java:431)
	at java.base/java.net.DatagramSocket.setOption(DatagramSocket.java:1207)
	at OptionsTest.test(OptionsTest.java:155)
	at OptionsTest.doMulticastSocketTests(OptionsTest.java:235)
	at OptionsTest.main(OptionsTest.java:337)
...
```
It works with the current master. Will try to figure it out.

BTW. did you try running jdk_net tests?

Comment (rdicroce) [reply src/java.base/windows/native/libnet/NetworkInterface.c]: I ran the tests for NetworkInterface specifically, though that reminds me that I need to open a bug report about fixpath mangling things incorrectly. I'll give the full jdk_net suite a run and see what happens.

Comment (djelinski) [reply src/java.base/windows/native/libnet/NetworkInterface.c]: ok so here goes: your patch changes the order in which the interfaces are returned; the original code returned them in order of increasing indexes, the new code appears to sort interfaces by LUID.
The failing test uses the first interface for which supportsMulticast returns true. I get `supportsMulticast` = `true` on all interfaces even with the old code.

The interface that happened to be on the first position of my list was:
```
67 name:ethernet_0 (Hyper-V Virtual Switch Extension Adapter-Hyper-V Virtual Switch Extension Filter-0000)
 ifup: true isP2P: false isLoopback: false macAddr: null MTU: 1500 supportsMulticast: true
```
it has no associated IP addresses, and is not usable with IP_MULTICAST_IF.
(Also it's a filter interface, which means that the old code returned ifUp = false).

Well that would explain how `supportsMulticast` is meant to be used; now we just need to figure out which interfaces can be used for multicasting.

Comment (rdicroce) [reply src/java.base/windows/native/libnet/NetworkInterface.c]: Note that the new code doesn't attempt to sort anything. If it happens to be sorted by LUID, that's the underlying API at work. The MS docs don't say anything about ordering, so this is probably an implementation detail and I wouldn't assume it will always be true.

I'm still running jdk_net, but I can already see that test failing on my machine. Though it has a different error:

```
java.io.IOException: Network interface not configured for IPv4
	at java.base/sun.nio.ch.DatagramChannelImpl.setOption(DatagramChannelImpl.java:389)
	at java.base/sun.nio.ch.DatagramSocketAdaptor.setOption(DatagramSocketAdaptor.java:431)
	at java.base/java.net.DatagramSocket.setOption(DatagramSocket.java:1207)
	at OptionsTest.test(OptionsTest.java:155)
	at OptionsTest.doMulticastSocketTests(OptionsTest.java:235)
	at OptionsTest.main(OptionsTest.java:337)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at com.sun.javatest.regtest.agent.MainWrapper$MainThread.run(MainWrapper.java:125)
	at java.base/java.lang.Thread.run(Thread.java:1623)
```

So I think when picking the interface to use for the test, we need to also check that isUp = true and that the list of IP addresses contains at least one IPv4 address.

Comment (dfuch) [reply src/java.base/windows/native/libnet/NetworkInterface.c]: Though it's more complicated than that. Some machines may have both IPv6 and IPv4, some machine may only have IPv4, and some machine may only have IPv6. I haven't looked at the test yet, but picking up the right interface on the various platforms has always been an issue for our tests.

**Comment by msheppar** [comment]
I don't think the NetworkInterface.h file should be dropped.
It is preferrable that typedef and struct decalarations remain in a header file rather than being merged into .c file.

File: src/java.base/windows/native/libnet/NetworkInterface.c
Comment (djelinski) [reply src/java.base/windows/native/libnet/NetworkInterface.c]: So,, apparently this is a Windows feature. [Older socket implementation](https://github.com/openjdk/jdk11/blob/37115c8ea4aff13a8148ee2b8832b20888a5d880/src/java.base/windows/native/libnet/TwoStacksPlainDatagramSocketImpl.c#L1350) had an explicit check for addresses bound to an interface, and only interfaces with associated addresses could be used for multicasting.

This needs to be fixed in the test. Can you add a check for addresses to `OptionsTest.getNetworkInterface`?

Beyond that, I'm fine with an implementation that unconditionally returns `true`.

File: src/java.base/windows/native/libnet/NetworkInterface.c
Comment (djelinski) [reply src/java.base/windows/native/libnet/NetworkInterface.c:73]: I tried creating an anycast address using the [CreateAnycastIpAddressEntry API](https://learn.microsoft.com/en-us/windows/win32/api/netioapi/nf-netioapi-createanycastipaddressentry) and the [New-NetIPAddress Powershell command](https://learn.microsoft.com/en-us/powershell/module/nettcpip/new-netipaddress?view=windowsserver2022-ps), but I only managed to lock myself out from a VM; creating the address failed with ERROR_NOT_SUPPORTED, and the powershell command additionally disabled DHCP on the selected interface.

With that in mind I'm fine with the code in its current form. We may revisit the issue if and when someone figures out how to add an anycast address.

File: src/java.base/windows/native/libnet/NetworkInterface.c
Comment (rdicroce) [reply src/java.base/windows/native/libnet/NetworkInterface.c]: Fixed.

File: src/java.base/windows/native/libnet/NetworkInterface.c
Comment (rdicroce) [reply src/java.base/windows/native/libnet/NetworkInterface.c]: Removed. Good catch.

File: src/java.base/windows/native/libnet/NetworkInterface.c
Comment (rdicroce) [reply src/java.base/windows/native/libnet/NetworkInterface.c]: Done.

File: src/java.base/windows/native/libnet/NetworkInterface.c
Comment (rdicroce) [reply src/java.base/windows/native/libnet/NetworkInterface.c]: Okay, I don't have any way to test it either, so leaving it as-is.

File: src/java.base/windows/native/libnet/NetworkInterface.c
Comment (rdicroce) [reply src/java.base/windows/native/libnet/NetworkInterface.c]: Done.

File: src/java.base/windows/native/libnet/NetworkInterface.c
Comment (rdicroce) [reply src/java.base/windows/native/libnet/NetworkInterface.c]: I agree, it looks like both of those would be fixed by these changes.

File: src/java.base/windows/native/libnet/NetworkInterface.c
Comment (rdicroce) [reply src/java.base/windows/native/libnet/NetworkInterface.c]: I've adjusted OptionsTest as @djelinski suggested, and changed the implementation to always return true. I left the giant comment so that when future devs wonder WTF we were thinking, they'll have the answer.

**Comment by rdicroce** [comment]
@msheppar I've moved the struct declaration back to the .h file.

The full jdk_net suite passes on my machine.

**Comment by djelinski** [comment]
I searched for other WinAPIs of interest; [GetIfStackTable](https://learn.microsoft.com/en-us/windows/win32/api/netioapi/nf-netioapi-getifstacktable) caught my attention.
As far as I could see, the network interfaces are arranged in a stack with many filter interfaces stacked on top of a single non-filter interface. All interfaces in a stack return the same values from `isUp`, `isPointToPoint`, `isLoopback` and `getHardwareAddress`. The filter interfaces do not have any IP addresses, and are not usable with multicast sockets.
I'm not sure how we could put that information to use. Ideas welcome.

**Comment by dfuch** [comment]
> I searched for other WinAPIs of interest; [GetIfStackTable](https://learn.microsoft.com/en-us/windows/win32/api/netioapi/nf-netioapi-getifstacktable) caught my attention.
> As far as I could see, the network interfaces are arranged in a stack with many filter interfaces stacked on top of a single non-filter interface. All interfaces in a stack return the same values from `isUp`, `isPointToPoint`, `isLoopback` and `getHardwareAddress`. The filter interfaces do not have any IP addresses, and are not usable with multicast sockets.
> I'm not sure how we could put that information to use. Ideas welcome.

Is there anything practical a Java Application could do with these interfaces? Except displaying them for debugging purposes? I mean - for instance - would it work if you tried, or are there case where it would be necessary,  to set one of these as the outgoing interface of a multicast socket?
Otherwise maybe we could just filter the filters out for now...

**Comment by rdicroce** [comment]
> > I searched for other WinAPIs of interest; [GetIfStackTable](https://learn.microsoft.com/en-us/windows/win32/api/netioapi/nf-netioapi-getifstacktable) caught my attention.
> > As far as I could see, the network interfaces are arranged in a stack with many filter interfaces stacked on top of a single non-filter interface. All interfaces in a stack return the same values from `isUp`, `isPointToPoint`, `isLoopback` and `getHardwareAddress`. The filter interfaces do not have any IP addresses, and are not usable with multicast sockets.
> > I'm not sure how we could put that information to use. Ideas welcome.
> 
> Is there anything practical a Java Application could do with these interfaces? Except displaying them for debugging purposes? I mean - for instance - would it work if you tried, or are there case where it would be necessary, to set one of these as the outgoing interface of a multicast socket? Otherwise maybe we could just filter the filters out for now...

I'm not sure if there's anything useful a Java application could do with the filter interfaces. I can't think of any. Having said that, the filter interfaces are returned by the existing code, so if we stop returning them, that is a functional change. I'm okay with that if you guys are.

But note that it is not guaranteed that all filters on top of an interface would return the same values for e.g. `isUp()`. The docs for [GetIfTable2Ex](https://learn.microsoft.com/en-us/windows/win32/api/netioapi/nf-netioapi-getiftable2ex) provide an example of this:

> An NDIS filter driver (LWF) can be attached on top of the NDIS intermediate driver. Assume that the NIC reports the MediaConnectState member of the MIB_IF_ROW2 structure as MediaConnectStateConnected but NDIS filter driver modifies the state and reports the state as MediaConnectStateDisconnected. When the interface information is queried with Level parameter set to MibIfTableNormal, the state at the top of the filter stack, that is MediaConnectStateDisconnected is reported. When the interface is queried with the Level parameter set to MibIfTableRaw, the state at the interface level directly, that is MediaConnectStateConnected is returned.

**Comment by rdicroce** [comment]
So, what's the verdict here? Do we keep the filter interfaces or get rid of them? If nobody has a strong opinion in favor of getting rid of them, I'm going to leave them alone, since filtering them out adds more complexity.

File: src/java.base/windows/native/libnet/NetworkInterface.h
Comment (AlanBateman) [inline src/java.base/windows/native/libnet/NetworkInterface.h:39]: Can NetworkInterface.h be dropped, and netaddr moved to NetworkInterface.c?

File: src/java.base/windows/native/libnet/NetworkInterface.c
Comment (AlanBateman) [inline src/java.base/windows/native/libnet/NetworkInterface.c:288]: I think it would be helpful for future maintainers if there was a comment on each of static functions to make it clear what they do.

File: src/java.base/windows/native/libnet/NetworkInterface.h
Comment (rdicroce) [reply src/java.base/windows/native/libnet/NetworkInterface.h:39]: I did that in an earlier revision. Then @msheppar asked for it to be moved back. See https://github.com/openjdk/jdk/pull/12593#issuecomment-1434450339

File: src/java.base/windows/native/libnet/NetworkInterface.c
Comment (AlanBateman) [inline src/java.base/windows/native/libnet/NetworkInterface.c:144]: Can you do a pass over this to handle the case where malloc fails?

File: src/java.base/windows/native/libnet/NetworkInterface.c
Comment (AlanBateman) [inline src/java.base/windows/native/libnet/NetworkInterface.c:646]: I don't think it's useful to include the historical note. Instead, I think reduce this down to a line or two to say that it optimistically assumes multicast is enabled.

**Comment by AlanBateman** [comment]
Overall I think this is a very contribution. It's good to modernize this code and remove the cruft that dates from before many of the new APIs were added.

I suspect we will need a release document any behaviour changes that may be observed when upgrading the JDK. In particular, the network interface names will look different. The synthesized names were names like "lo" and "eth0" whereas the new names come from Windows (names like "lookup_0" and "ethernet_0"). I haven't found tools that show the names that ConvertInterfaceLuidToNameW returns but if there are then the RN could mention it. On other platforms, it would be was easy for configuration to name the interface to be used for multicasting, it doesn't seem easy to do the same on Windows and application might have to resort to configuration by IP address.

I'm not too concerned that interfaces without IP addresses are enumerated. They aren't useful for multicasting applications but maybe they are useful for informational purposes, e.g. programs that collect system environment. I'm also not too concern about the ordering of the IP addresses. I'll defer to @djelinski on whether some JDK tests or test infra needs to be updated in advance to ensure that tests that choose interfaces will continue to work.

**Comment by djelinski** [comment]
Ok then, let's keep the filter interfaces as they are; we can always revisit the issue later if there's demand.
I'm already working on updating the JCK tests.

File: src/java.base/windows/native/libnet/NetworkInterface.c
Comment (rdicroce) [reply src/java.base/windows/native/libnet/NetworkInterface.c:295]: Since there were no cases left where throwIfNotFound was true, I have removed it altogether.

**Comment by rdicroce** [comment]
I've updated the PR to address all of the various comments.

@AlanBateman Regarding tools that return the new interface names, I'm not aware of any either. But it's not too hard to hack together a PowerShell script to call if_indextoname which does exactly what the name implies.

File: src/java.base/windows/native/libnet/ResolverConfigurationImpl.c
Comment (djelinski) [inline src/java.base/windows/native/libnet/ResolverConfigurationImpl.c:190]: Is this needed? As far as I could tell, we only retrieve DNS addresses here, and we only want the DNS servers associated with usable interfaces.

**Review by djelinski — COMMENTED** [review]
Network interface changes LGTM now. I'm not sure about the change to ResolverConfiguration, see below.

File: src/java.base/windows/native/libnet/ResolverConfigurationImpl.c
Comment (djelinski) [reply src/java.base/windows/native/libnet/ResolverConfigurationImpl.c:190]: Yes I think so. This change doesn't provide any obvious benefit, so let's revert it.

Comment (rdicroce) [reply src/java.base/windows/native/libnet/ResolverConfigurationImpl.c]: Done.

**Comment by djelinski** [comment]
Thanks for that!
I'm still running one last round of tests and will approve this if there are no new failures.

I'll start working on a CSR and a release note.

/csr needed

**Review by djelinski — APPROVED** [review]

**Comment by djelinski** [comment]
CSR approved, release note ready. I think we're good to go. @rdicroce you can integrate the changes now.

**Review by Michael-Mc-Mahon — APPROVED** [review]

**Pull request closed without merging**
